src/org/etools/j1939_84/bus/j1939/packets/Slot.java

```diff
@@ -19,197 +18,318 @@
-	 *
-	 * @param value
-	 *              the value to scale
-	 * @return double
-	 */
-	public double scale(double value) {
-		return value * getScaling() + getOffset();
-	}
+    /**
+     * The Map of SLOT ID to Slot for lookups
+     */
+    private static Map<Integer, Slot> slots;
+
+    /**
+     * Helper method to convert from the given {@link String} to a double
+     *
+     * @param string the {@link String} to convert
+     * @return Double or null if the string cannot be converted
+     */
+    private static Double doubleValue(String string) {
+        try {
+            return Double.parseDouble(string);
+        } catch (NumberFormatException e) {
+            return null;
+        }
+    }
+
+    /**
+     * Finds a SLOT given the Identifier of the SLOT
+     *
+     * @param id the Identifier of the SLOT
+     * @return a SLOT or null if not found
+     */
+    public static Slot findSlot(int id) {
+        return getSlots().get(id);
+    }
+
+    /**
+     * Caches and returns all known {@link Slot}s
+     *
+     * @return a Map of SLOT ID to Slot
+     */
+    private static Map<Integer, Slot> getSlots() {
+        if (slots == null) {
+            slots = loadSlots();
+
+        }
+        return slots;
+    }
+
+    /**
+     * Helper method to convert from the given {@link String} to an int
+     *
+     * @param string       the {@link String} to convert
+     * @param defaultValue the int value if the string cannot be converted
+     * @return int
+     */
+    private static int intValue(String string, int defaultValue) {
+        try {
+            return Integer.parseInt(string);
+        } catch (NumberFormatException e) {
+            return defaultValue;
+        }
+    }
+
+    /**
+     * Read the slots.csv file which contains all the SLOTs
+     *
+     * @return Map of SLOT ID to Slot
+     */
+    private static Map<Integer, Slot> loadSlots() {
+        Map<Integer, Slot> slots = new HashMap<>();
+        String[] values;
+
+        final InputStream is = Resources.class.getResourceAsStream("slots.csv");
+        final InputStreamReader isReader = new InputStreamReader(is, StandardCharsets.ISO_8859_1);
+        try (CSVReader reader = new CSVReader(isReader)) {
+            while ((values = reader.readNext()) != null) {
+                final int id = Integer.parseInt(values[0]);
+                final String name = values[1];
+                final String type = values[2];
+                final Double scaling = doubleValue(values[3]);
+                final String unit = values[4];
+                final Double offset = doubleValue(values[5]);
+                final int length = intValue(values[6], -1);
+                Slot slot = new Slot(id, name, type, scaling, offset, unit, length);
+                slots.put(id, slot);
+            }
+        } catch (Exception e) {
+            J1939_84.getLogger().log(Level.SEVERE, "Error loading map from slots", e);
```

Comment:
I think a failure to load the file should be a runtime exception, not just a log message.

Reply:
Runtime exception would cause the app to crash.

Reply:
If we don't crash the app, then the output may be wrong.  The user should be notified that the results will be invalid.